scripts/recursive_build.py

```diff
@@ -122,206 +88,597 @@ def run_cmd(cmd):
-        if json_file_path and os.path.exists(json_file_path):
-            with open(json_file_path, 'r') as f:
-                data = json.load(f)
-                for obj in data:
-                    if name in obj:
-                        mcus = obj[name].get('mcus', [])
-                        toolchain = obj[name].get('toolchain', '')
-
-                        if architecture not in toolchain_mcu_map:
-                            toolchain_mcu_map[architecture] = {}
-
-                        compilers = compiler_list.get(architecture.upper(), [])
-                        for compiler in compilers:
-                            if toolchain == 'gcc_clang':
-                                if 'gcc' in compiler or 'clang' in compiler:
-                                    if compiler in toolchain_mcu_map[architecture]:
-                                        toolchain_mcu_map[architecture][compiler].extend(mcus)
-                                    else:
-                                        toolchain_mcu_map[architecture][compiler] = mcus
-                            else:
-                                if toolchain in compiler:
-                                    if compiler in toolchain_mcu_map[architecture]:
-                                        toolchain_mcu_map[architecture][compiler].extend(mcus)
-                                    else:
-                                        toolchain_mcu_map[architecture][compiler] = mcus
-
-    # Remove duplicates from each list in the map.
-    for architecture in toolchain_mcu_map:
-        for toolchain in toolchain_mcu_map[architecture]:
-            toolchain_mcu_map[architecture][toolchain] = list(set(toolchain_mcu_map[architecture][toolchain]))
-
-    return toolchain_mcu_map
+# Runs the build commands for each member of mcu_list, board_list, and mcu_card_list.
+def run_builds(changes_dict):
+    # Get the SDK version from manifest.json file.
+    sdk_version = get_sdk_version()
 
-def main():
-    global build_failed
+    # Run build for all MCUs from mcu_list.
+    print(f"\033[93mRunning build for {len(changes_dict['mcu_list'])} MCUs\033[0m")
+    for mcu in changes_dict['mcu_list']:
+        # Get the necessary compiler for the current MCU build.
+        compilers = get_compilers(mcu, is_mcu=True)
+        for compiler in compilers:
+            cmd = f'xvfb-run --auto-servernum --server-num=1 {toolPath} --isBareMetal "1" --compiler "{compiler}" --sdk "{sdk_version}" --board "GENERIC_ARM_BOARD" --mcu "{mcu}" --installPrefix "{testPath}/mcu_build/{compiler}"'
+            run_cmd(cmd, changes_dict, mcu + ' ' + compiler)
 
-    # Dictionary for architecture-compiler-mcu dependencies
-    compiler_mcu_map = {}
+# Returns the list of compilers based on the given name and type.
+def get_compilers(name, is_mcu=True):
+    if is_mcu:
+        if any(substring in name for substring in ["ATSAM", "STM", "TM4C", "MK"]):
```

Comment:
Check if we have any `SAM` MCUs.
I believe not all ATMEL 32-bit ones start with `ATSAM`.